### ISSUE TITLE
sdk/sync: use a custom `Dialer` to cache redis host

### ISSUE THREAD

**Issue by raulk** [testground/testground#1054]
## Context

Prior to https://github.com/testground/sdk-go/pull/18, we used to resolve Redis hostname and use the IP address in the client pool configuration.

While this might be OK for short-lived processes (test plans), it hinders resilience of long-lived processes (sidecars), because Redis might fail and be restarted on a different IP _during_ the lifetime of the process. Since the Redis pool is bound to the old IP, there is no chance of recovery whatsoever.

However, the cited PR introduced a likely regression in that using the hostname and forcing a resolution (almost) every time we create a connection does not scale well. 10k instances x 2-3 Redis connections each could give way to a 30k DNS query storm. That's precisely why we preempted the resolution.

## Proposal

Use the `redis.Options#Dialer` field to set a custom dialer that keeps an IP address cache, and only falls back dialing the hostname if the IP dial fails.

**Comment by nonsense** [comment]
An alternative solutions would be to install Kubernetes NodeLocal DNS cache, so that we have a cache across all test plan instances in on a single host, and not a cache per test plan instance - when running 10k pods, usually we have more than 60 instances per host.

**Comment by nonsense** [comment]
Benchmarks on CoreDNS, just to have some intuition on what we should expect from CoreDNS on k8s - https://github.com/coredns/deployment/blob/master/kubernetes/Scaling_CoreDNS.md

**Comment by raulk** [comment]
IIRC, the simple solution we discussed off-band to (a) avoid stale entries and (b) remove a time-based TTL from the equation, was to resolve the Redis host once per test run ID. All subsequent containers scheduled in the same host for the same test run ID would reuse the resolved IP address. Once all containers for that test run are stopped, we erase the entry in the cache.

Correct, @nonsense?

**Comment by nonsense** [comment]
Yes, that's right. The assumption is that Redis is not persisted, so if it is restarted (during a release or re-deploy on TaaS for example), the test runs during that window will crash anyways.
